### PR TITLE
implemented changes for method & event Triggers refactoring

### PR DESCRIPTION
**Pull request by warbhe-amit**
Implemented changes for method & event Triggers refactoring

### CODE REVIEW

**Review by michaelcfine — CHANGES_REQUESTED** [review]
This seems to be code to implement the 'broadcastEvent' feature, not really refactoring triggers, but that's okay. This looks pretty good, but I think in all five cases where ctx objects are being created, you're not calling the right function with the right signature for sendEvent and sendBroadcastEvent. See my comments. In all cases, ctx.sendEvent should call events.sendEvent(<no boolean param>) and ctx.sendBroadcastEvent should call events.sendBroadcastEvent(<no boolean param>). The refactoring to coreSendEvent should have left the signatures for those methods NOT having a boolean param; only coreSendEvent has the boolean and it is private; so calling code should not pass it.

Also... See my comment about refactoring the events.coreSendEvent function when it comes to sending messages over the sockets. Please refactor so that first you get a list of sockets (different logic depending on broadcast or not) and then loop through the list (which can now be 100% the same code).

Also... please test thoroughly. I don't see how this code would have run properly with the invocations not matching the signatures. Thanks.

File: server/src/events.mjs [hunk 1]
```diff
@@ -38,6 +38,23 @@ function registerEventListener(userId, oMsg) {
   logger.debug(`Registered event listener mapping: ${userId}:${oMsg.method}:${oMsg.id}`);
 }
 
+// Return true if at least one user in the user’s group is registered for the given method
+function isAnyRegisteredInGroup(userId, method) {
+  const userList = userManagement.getUserListForUser(userId);
+  for (const user of userList){
+    console.log('user', user);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/events.mjs:45]: Done

File: server/src/events.mjs [hunk 2]
```diff
@@ -38,6 +38,23 @@ function registerEventListener(userId, oMsg) {
...
+    if ( isRegisteredEventListener(user, method) ){
+      return true;
+    }
+  }
+  return false;
+}
+
+// Check the user’s group is registered for the given method
+function isRegisteredEventListenerInGroup(userId, method) {
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/events.mjs:54]: Done

File: server/src/events.mjs [hunk 3]
```diff
@@ -129,7 +159,19 @@ function sendEvent(ws, userId, method, result, msg, fSuccess, fErr, fFatalErr) {
                 function fFatalErr() {
                   logger.info(`Internal error`)
                 }
-                sendEvent(ws, userId, method, result, msg, fSuccess, fErr, fFatalErr);
+                sendEvent(false, ws, userId, method, result, msg, fSuccess, fErr, fFatalErr);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/events.mjs:162]: Done

File: server/src/events.mjs [hunk 4]
```diff
@@ -129,7 +159,19 @@ function sendEvent(ws, userId, method, result, msg, fSuccess, fErr, fFatalErr) {
...
+                function fSuccess() {
+                  logger.info(`${msg}: Sent event ${onMethod} with result ${JSON.stringify(result)}`)
+                }
+                function fErr() {
+                  logger.info(`Could not send ${onMethod} event because no listener is active`)
+                }
+                function fFatalErr() {
+                  logger.info(`Internal error`)
+                }
+                events.sendEvent(true, ws, userId, onMethod, result, msg, fSuccess, fErr, fFatalErr);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/events.mjs:174]: Done

File: server/src/events.mjs [hunk 5]
```diff
@@ -163,7 +205,19 @@ function sendEvent(ws, userId, method, result, msg, fSuccess, fErr, fFatalErr) {
                 function fFatalErr() {
                   logger.info(`Internal error`)
                 }
-                sendEvent(ws, method, result, msg, fSuccess, fErr, fFatalErr);
+                sendEvent(false, ws, method, result, msg, fSuccess, fErr, fFatalErr);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/events.mjs:208]: Done

File: server/src/messageHandler.mjs [hunk 6]
```diff
@@ -124,7 +124,19 @@ async function handleMessage(message, userId, ws) {
             function fFatalErr() {
               logger.info(`Internal error`)
             }
-            events.sendEvent(ws, userId, onMethod, result, msg, fSuccess, fErr, fFatalErr);
+            events.sendEvent(false, ws, userId, onMethod, result, msg, fSuccess, fErr, fFatalErr);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/messageHandler.mjs:127]: Done

File: server/src/messageHandler.mjs [hunk 7]
```diff
@@ -124,7 +124,19 @@ async function handleMessage(message, userId, ws) {
...
+            function fSuccess() {
+              logger.info(`${msg}: Sent event ${onMethod} with result ${JSON.stringify(result)}`)
+            }
+            function fErr() {
+              logger.info(`Could not send ${onMethod} event because no listener is active`)
+            }
+            function fFatalErr() {
+              logger.info(`Internal error`)
+            }
+            events.sendEvent(true, ws, userId, onMethod, result, msg, fSuccess, fErr, fFatalErr);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/messageHandler.mjs:139]: Done

File: server/src/messageHandler.mjs [hunk 8]
```diff
@@ -173,7 +185,19 @@ async function handleMessage(message, userId, ws) {
...
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/messageHandler.mjs:200]: Done

File: server/src/stateManagement.mjs [hunk 9]
```diff
@@ -151,7 +151,19 @@ function handleDynamicResponseValues(userId, methodName, params, ws, resp){
...
+          function fSuccess() {
+            logger.info(`${msg}: Sent event ${onMethod} with result ${JSON.stringify(result)}`)
+          }
+          function fErr() {
+            logger.info(`Could not send ${onMethod} event because no listener is active`)
+          }
+          function fFatalErr() {
+            logger.info(`Internal error`)
+          }
+          events.sendEvent(true, ws, userId, onMethod, result, msg, fSuccess, fErr, fFatalErr);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/stateManagement.mjs:166]: Done

File: server/src/stateManagement.mjs [hunk 10]
```diff
@@ -151,7 +151,19 @@ function handleDynamicResponseValues(userId, methodName, params, ws, resp){
           function fFatalErr() {
             logger.info(`Internal error`)
           }
-          events.sendEvent(ws, userId, onMethod, result, msg, fSuccess, fErr, fFatalErr);
+          events.sendEvent(false, ws, userId, onMethod, result, msg, fSuccess, fErr, fFatalErr);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/stateManagement.mjs:154]: Done

File: server/src/events.mjs [hunk 11]
```diff
@@ -163,7 +205,19 @@ function sendEvent(ws, userId, method, result, msg, fSuccess, fErr, fFatalErr) {
...
+                function fSuccess() {
+                  logger.info(`${msg}: Sent event ${onMethod} with result ${JSON.stringify(result)}`)
+                }
+                function fErr() {
+                  logger.info(`Could not send ${onMethod} event because no listener is active`)
+                }
+                function fFatalErr() {
+                  logger.info(`Internal error`)
+                }
+                events.sendEvent(true, ws, userId, onMethod, result, msg, fSuccess, fErr, fFatalErr);
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/events.mjs:220]: Done

File: server/src/events.mjs [hunk 12]
```diff
@@ -184,25 +238,40 @@ function sendEvent(ws, userId, method, result, msg, fSuccess, fErr, fFatalErr) {
...
-          const oEventMessage = {
-            jsonrpc: '2.0',
-            id: id,
-            result: finalResult
-          };
-          const eventMessage = JSON.stringify(oEventMessage);
-          // Could do, but why?: const dly = stateManagement.getAppropriateDelay(user, method); await util.delay(dly);
-          ww.send(eventMessage);
-          logger.info(`${msg}: Sent event message: ${eventMessage}`);
+      let id;
```
[the remark this answers is not in the document]
Comment (warbhe-amit) [reply server/src/events.mjs]: done

**Review by michaelcfine — APPROVED** [review]
👍 Looks good.

**Pull request merged**